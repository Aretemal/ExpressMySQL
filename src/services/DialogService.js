import { Op } from 'sequelize';
import Dialog from '../models/dialog.js';
import Message from '../models/message.js';
import User from '../models/user.js';

class DialogService {
  async sendMessage({ message }, recipientId, senderId) {
    let dialog = await Dialog.findOne({
      where: {
        [Op.or]: [
          { firstId: senderId, secondId: recipientId },
          { secondId: senderId, firstId: recipientId },
        ],
      },
    });
    if (!dialog) {
      dialog = await Dialog.create({ firstId: senderId, secondId: recipientId });
    }
    const content = await Message.create({ message, dialogId: dialog.id, senderId });
    return content;
  }

  async getAllMessage(id) {
    const messages = await Message.findAll({
      where: {
        dialogId: id,
      },
    });
    return messages;
  }

  async getAllDialogs(id) {
    const dialogsId = await Dialog.findAll({
      attributes: ['id', 'firstId', 'secondId'],
      where: {
        [Op.or]: [
          { firstId: id },
          { secondId: id },
        ],
      },
      group: ['firstId', 'secondId'],
    });
    const usersId = dialogsId.map((dialog) => {
      if (dialog.firstId === id) {
        return dialog.secondId;
      }
      return dialog.firstId;
    });
    const uniqueId = [...new Set(usersId)];
    const data = await User.findAll({
      attributes: ['id', 'firstName', 'lastName', 'login', 'email', 'status', 'ava'],
      where: {
        id: {
          [Op.in]: uniqueId,
        },
      },
    });
    const users = data.map((user) => user.dataValues);
    for (let i = 0; i < users.length; i++) {
      users[i].dialogId = dialogsId[i].id;
    }
    return users;
  }
}

export default new DialogService();
