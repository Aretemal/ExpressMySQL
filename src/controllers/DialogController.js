import { validationResult } from 'express-validator';
import CollectionSerializer from '../serializers/CollectionSerializer.js';
import DialogSerializer from '../serializers/DialogSerializer.js';
import MessageSerializer from '../serializers/MessageSerializer.js';
import DialogService from '../services/DialogService.js';
import { getClass } from '../utils/getClass.js';

class DialogController {
  async sendMessage(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const message = await DialogService.sendMessage(req.body, req.params.id, req.user.id);
    req.serializer = new MessageSerializer(message);
    next();
  }

  async getAllDialogs(req, res, next) {
    const users = await DialogService.getAllDialogs(req.user.id);
    req.serializer = new CollectionSerializer(users, { serializerType: DialogSerializer });
    next();
  }

  async getAllMessage(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const messages = await DialogService.getAllMessage(req.params.id);
    req.serializer = new CollectionSerializer(
      messages,
      {
        serializerType: MessageSerializer,
        include: ['user'],
        getter: getClass,
      },
    );
    next();
  }
}
export default new DialogController();
