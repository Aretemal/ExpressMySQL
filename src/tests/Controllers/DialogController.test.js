/* eslint-disable no-undef */
import { jest } from '@jest/globals';
import DialogController from '../../controllers/DialogController';
import DialogService from '../../services/DialogService';

jest.fn('DialogService');
const req = {
  body: {
    message: 'Hello',
    recipientId: 5,
  },
  user: {
    id: 1,
  },
  params: {
    id: 1,
  },
};
const res = {
  dataJS: null,
  json: (data) => {
    res.dataJS = data;
  },
};

describe('Dialog Controller :', () => {
  describe('sendMessage', () => {
    test('should create and return new message', async () => {
      function next() {}
      DialogService.sendMessage = () => ({
        message: 'message',
        dialogId: '1',
        senderId: '2',
      });

      await DialogController.sendMessage(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });

  describe('deleteMessage', () => {
    test('should return deleted message', async () => {
      function next() {}
      DialogService.deleteMessage = () => ({
        message: 'message',
        dialogId: '1',
        senderId: '2',
      });

      await DialogController.deleteMessage(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });

  describe('getAllDialogs', () => {
    test('should return dialog array', async () => {
      function next() {}
      DialogService.getAllDialogs = () => ([{
        name: 'name',
        dialogId: '1',
        id: '2',
      }]);

      await DialogController.getAllDialogs(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });

  describe('getAllMessage', () => {
    test('should return message array', async () => {
      function next() {}
      DialogService.getAllMessage = () => ([{
        message: 'message',
        dialogId: '1',
        senderId: '2',
      }]);

      await DialogController.getAllMessage(req, res, next);

      expect(req.serializer).toMatchSnapshot();
    });
  });
});
