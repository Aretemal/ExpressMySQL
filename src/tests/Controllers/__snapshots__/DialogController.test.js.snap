// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Controller : deleteMessage should return deleted message 1`] = `
MessageSerializer {
  "body": {
    "message": "Hello",
    "recipientId": 5,
  },
  "params": {
    "id": 1,
  },
  "resource": {
    "dialogId": "1",
    "message": "message",
    "senderId": "2",
  },
  "serializer": MessageSerializer {
    "resource": {
      "dialogId": "1",
      "message": "message",
      "senderId": "2",
    },
  },
  "user": {
    "id": 1,
  },
}
`;

exports[`Dialog Controller : getAllDialogs should return dialog array 1`] = `
CollectionSerializer {
  "resource": [
    {
      "dialogId": "1",
      "id": "2",
      "name": "name",
    },
  ],
  "serializerType": [Function],
}
`;

exports[`Dialog Controller : getAllMessage should return message array 1`] = `
CollectionSerializer {
  "getter": [Function],
  "include": [
    "user",
  ],
  "resource": [
    {
      "dialogId": "1",
      "message": "message",
      "senderId": "2",
    },
  ],
  "serializerType": [Function],
}
`;

exports[`Dialog Controller : sendMessage should create and return new message 1`] = `
MessageSerializer {
  "resource": {
    "dialogId": "1",
    "message": "message",
    "senderId": "2",
  },
}
`;
